mm2src/mm2_metrics/src/adapt.rs

```diff
@@ -0,0 +1,381 @@
+            let key = $crate::metrics::Key::from_parts(key_name, from_slice_to_labels(&[($($label_key, $label_val),+)]));
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+         }
+    }};
+}
+
+#[allow(dead_code)]
+fn from_slice_to_labels(labels: &[(&'static str, &'static str)]) -> Vec<Label> { labels.into_labels() }
+
+pub struct Snapshot {
+    pub counters: HashMap<String, HashMap<Vec<String>, u64>>,
+    pub gauges: HashMap<String, HashMap<Vec<String>, u64>>,
+    pub histograms: HashMap<String, HashMap<Vec<String>, f64>>,
+}
+
+#[allow(dead_code)]
+pub(crate) struct Inner {
+    pub registry: Registry<metrics::Key, GenerationalAtomicStorage>,
+}
+
+impl Inner {
+    fn get_metrics(&self) -> Snapshot {
+        let mut counters = HashMap::new();
+        for (key, counter) in self.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            let entry = counters
+                .entry(key.to_string())
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        let mut gauges = HashMap::new();
+        for (key, gauge) in self.registry.get_gauge_handles() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            let entry = gauges
+                .entry(key.to_string())
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        let mut histograms = HashMap::new();
+        let histogram_handles = self.registry.get_histogram_handles();
+        for (key, histogram) in histogram_handles {
+            let (name, labels) = key_to_parts(&key, None);
+            let value = histogram.get_inner().data().iter().sum();
+            let entry = histograms
+                .entry(name)
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
+        let mut output: Vec<PreparedMetric> = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles().drain() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
```

Comment:
Please try to add a generic function here as well

Reply:
done